Fixed tests in distributed task

IzabellaRaulin opened the pull request:
Fixes the following failures:

![image](https://cloud.githubusercontent.com/assets/11335874/24364177/3a60246a-1312-11e7-9613-dc1ac4aac9f9.png)


### Summary of changes:
- unsubscribe plugins before emitting of TaskEnded(or TaskDisabled) event

Before  |  After
-------- |---------
change state to core.TaskEnded |   change state to core.TaskEnded
emit TaskEndedEvent  |   unsubscribePlugins
unsubscribePlugins  |  emit TaskEndedEvent

Testing done:
- still in progress

@intelsdi-x/snap-maintainers

IzabellaRaulin committed:
Improved error messaging for `plugin unload`
Changed the error message to more appropriate one for `plugin swap` when ver < 1

IzabellaRaulin commented on scheduler/scheduler.go:
```diff
@@ -761,9 +761,6 @@ func (s *scheduler) HandleGomitEvent(e gomit.Event) {
 		}).Debug("event received")
-		// We need to unsubscribe from deps when a task has ended
-		task, _ := s.getTask(v.TaskID)
-		task.UnsubscribePlugins()
```
FYI:
This happens before emitting the event - see [scheduler/task.go](https://github.com/IzabellaRaulin/snap/pull/10/files#diff-77f163e381dd113e90ebd1bd76479b54R492)

IzabellaRaulin commented on scheduler/scheduler.go:
```diff
@@ -773,9 +770,6 @@ func (s *scheduler) HandleGomitEvent(e gomit.Event) {
 		}).Debug("event received")
-		// We need to unsubscribe from deps when a task goes disabled
-		task, _ := s.getTask(v.TaskID)
-		task.UnsubscribePlugins()
```
FYI:
This happens before emitting the event - see [scheduler/task.go](https://github.com/IzabellaRaulin/snap/pull/10/files#diff-77f163e381dd113e90ebd1bd76479b54R499)

IzabellaRaulin commented on scheduler/task.go:
```diff
@@ -321,37 +330,24 @@ func (t *task) stream() {
-		return ErrTaskDisabledOnFailures
-	}
-	return nil
-}
```
FYI:
I see the reasoning why Cody added this `checkTaskFailures()` to avoid duplicating the same code, but the error log loses its value because this function might be called everywhere, so "_block" does not point to the exact place. So, I changed it to the following:
```
if t.stopOnFailure >= 0 && consecutiveFailures >= t.stopOnFailure {
		taskLogger.WithFields(log.Fields{
			"_block":               "stream",
			"task-id":              t.id,
			"task-name":            t.name,
			"consecutive failures": consecutiveFailures,
			"error":                t.lastFailureMessage,
		}).Error(ErrTaskDisabledOnFailures)
		// disable the task
		t.disable(t.lastFailureMessage)
		return
}
```

IzabellaRaulin commented on scheduler/task.go:
```diff
@@ -537,6 +527,36 @@ func (t *task) fire() {
+
+	// We need to unsubscribe from deps when a task has disabled
+	t.UnsubscribePlugins()
+}
```
FYI:
The only my inaccuracy is about handling errors from `UnsubscribePlugins()`.  I would like to add here log.Error at least. What do You think about that?

Notice that even in historical commit https://github.com/intelsdi-x/snap/blob/c592966ef49f7c1dbb78b712a6d3f802a7e00b8e/scheduler/scheduler.go#L806 there was no react if unsubscribing failed.

jcooklin replied:
> would like to add here log.Error at least. What do You think about that?

I agree.  Let's do it.

IzabellaRaulin replied:
I added some logs inside [UnsubcribePlugins()](https://github.com/IzabellaRaulin/snap/pull/10/files#diff-77f163e381dd113e90ebd1bd76479b54R375)

janczer committed:
Add verbs to error message

IzabellaRaulin committed:
Provided snaptel command usage after error message

jcooklin reviewed (approved):
LGTM

candysmurf committed:
Merge pull request #1572 from janczer/master

Add verb to error message

IzabellaRaulin committed:
Merge pull request #1571 from IzabellaRaulin/unload_plg_ver

Improved error message for `plugin unload` cmd

Commit:
Update CONTRIBUTING.md

Title is now properly formatted.

IzabellaRaulin committed:
Merge pull request #1581 from skonefal/patch-1

Fixed title in CONTRIBUTING.md

sayantanigoswami committed:
Fix: #1578 Fix typo in snapteld configuration file

To auto-load plugins and tasks, a colon separated list of directories
must be provided. This fixes the typo that states it should be a comma
separated list of directories.

rashmigottipati committed:
Merge pull request #1583 from sayantanigoswami/sayantani/fix-typo

Fix: #1578 Fix typo in snapteld configuration file

ami-GS committed:
fix code area appearance

candysmurf committed:
Merge pull request #1585 from ami-GS/devel3

fix code area appearance

jcooklin committed:
Fixes #1576 - flakey tests scheduler

- Adds event for when plugins are unsubscribed
- Listens for unsubscription event before asserting

IzabellaRaulin committed:
Fixed flaky test in windowed schedule and added method to proceed task disabling
